packages/cojson-transport-ws/src/index.ts

```diff
@@ -1,108 +1,121 @@
+            Stream.fromEffect,
+        );
 
-            pingTimeout = setTimeout(() => {
-                console.debug("Ping timeout");
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.left(new PingTimeoutError())),
-                );
-                try {
-                    ws.close();
-                } catch (e) {
-                    console.error(
-                        "Error while trying to close ws on ping timeout",
-                        e,
-                    );
-                }
-            }, 2500);
+        const isSyncMessage = (msg: unknown): msg is SyncMessage => {
+            // boo side-effects
 
-            if (msg.type === "ping") {
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings =
-                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                    (globalThis as any).jazzPings || [];
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings.push({
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            if ((msg as any)?.type === "ping") {
+                const ping = msg as PingMsg;
+                g.jazzPings ||= [];
+                g.jazzPings.push({
                     received: Date.now(),
-                    sent: msg.time,
-                    dc: msg.dc,
+                    sent: ping.time,
+                    dc: ping.dc,
                 });
-                return;
-            } else {
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.right(msg)),
-                );
+                return false;
             }
-        });
+            return true;
+        };
 
-        ws.addEventListener("open", () => {
-            void Stream.fromQueue(outgoing).pipe(
-                Stream.runForEach((msg) =>
-                    Effect.sync(() => ws.send(JSON.stringify(msg))),
+        yield* Queue.take(outgoing).pipe(
+            Effect.andThen((message) => ws.send(JSON.stringify(message))),
+            Effect.forever,
+            Effect.whenEffect(
+                fromEventListener(ws, "open").pipe(Effect.as(true)),
+            ),
+            Effect.forkDaemon, // :/
+        );
+
+        type E = WebsocketEvents["message"];
+        const messages = Stream.fromEventListener<E>(ws_, "message").pipe(
+            Stream.timeoutFail(() => new PingTimeoutError(), "2.5 seconds"),
+            Stream.tapError((_e) =>
+                Console.warn("Ping timeout").pipe(
+                    Effect.zipRight(Effect.try(() => ws.close())),
+                    Effect.catchAll((e) =>
+                        Console.error(
+                            "Error while trying to close ws on ping timeout",
+                            e,
+                        ),
+                    ),
```

Comment:
can we consider gen here too to avoid the jargon?

```ts
Effect.gen(function*() {
  yield* Console.warn("Ping timeout");
  yield* Effect.try(() => ws.close).pipe(
    Effect.catchAll(e => Console.error(...)) // really considering try/catch here
  )
});

Reply:
I can use `.andThen` here as well instead of `zipRight`. I prefer the pipe version because of how aggressive the indents in this project are haha, tell me what you think

Reply:
I think I'm going to switch to 2 spaces soon...